### PR TITLE
[FLASK] Add signature insights

### PR DESCRIPTION
**Pull request by hmalik88**
Per [SIP-16](https://github.com/MetaMask/SIPs/blob/main/SIPS/sip-16.md), the signature insights implementation is being integrated.

Closes https://github.com/MetaMask/MetaMask-planning/issues/1778

Demo: 

https://github.com/MetaMask/metamask-extension/assets/41640681/a238e720-c614-4be8-b272-5257c1ccfad2

### CODE REVIEW

File: shared/constants/snaps/permissions.ts [hunk 1]
```diff
@@ -8,6 +8,7 @@ export const EndowmentPermissions = Object.freeze({
   'endowment:lifecycle-hooks': 'endowment:lifecycle-hooks',
   ///: BEGIN:ONLY_INCLUDE_IF(build-flask)
   'endowment:page-home': 'endowment:page-home',
+  'endowment:signature-insight': 'endowment:signature-insight',
```
Comment (FrederikBolding) [inline shared/constants/snaps/permissions.ts:11]: If we are activating this permission it needs to specify a permission name, icon etc

File: ui/components/app/signature-request-siwe/signature-request-siwe.js [hunk 2]
```diff
@@ -172,124 +185,163 @@ export default function SignatureRequestSIWE({ txData }) {
   const rejectNText = t('rejectRequestsN', [messagesCount]);
```
Comment (FrederikBolding) [inline ui/components/app/signature-request-siwe/signature-request-siwe.js:186]: These changes seem complex. Why did it end up requiring this many changes?

File: ui/hooks/useSignatureInsights.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  useEffect(() => {
+    let cancelled = false;
+
+    async function fetchInsight() {
+      setLoading(true);
+
+      const { msgParams } = txData;
+      const signature = {
+        from: msgParams.from,
+        data: JSON.parse(JSON.stringify(msgParams.data)),
```
Comment (FrederikBolding) [inline ui/hooks/useSignatureInsights.js:27]: This seems flawed if `msgParams.data` is a string. I wonder if we should use the sanitized data object from the UI 🤔

File: ui/pages/confirm-signature-request/index.js [hunk 4]
```diff
@@ -74,6 +80,52 @@ const ConfirmTxScreen = ({ match }) => {
   const [prevValue, setPrevValues] = useState();
   const history = useHistory();
 
+  const getTxData = useCallback(() => {
```
Comment (FrederikBolding) [inline ui/pages/confirm-signature-request/index.js:83]: Why was this moved?

**Review by FrederikBolding — CHANGES_REQUESTED** [review]

File: ui/hooks/useSignatureInsights.js (hunk 3 above)
Comment (hmalik88) [reply ui/hooks/useSignatureInsights.js]: Updated

File: ui/pages/confirm-signature-request/index.js (hunk 4 above)
Comment (hmalik88) [reply ui/pages/confirm-signature-request/index.js:83]: `txData` is not defined otherwise

File: shared/constants/snaps/permissions.ts (hunk 1 above)
Comment (hmalik88) [reply shared/constants/snaps/permissions.ts:11]: Yeah, going to work on this part with @eriknson

Comment (hmalik88) [reply shared/constants/snaps/permissions.ts:11]: Done.

File: shared/constants/snaps/permissions.ts [hunk 5]
```diff
@@ -22,8 +23,6 @@ export const ExcludedSnapPermissions = Object.freeze({
 });
 
 export const ExcludedSnapEndowments = Object.freeze({
-  'endowment:signature-insight':
```
Comment (FrederikBolding) [inline shared/constants/snaps/permissions.ts:25]: Should be moved to the below fence to be disallowed in main.

File: ui/pages/confirm-signature-request/index.js (hunk 4 above)
Comment (FrederikBolding) [reply ui/pages/confirm-signature-request/index.js:83]: I don't understand. Couldn't we just add the insight hook after where txData was defined before?

File: ui/hooks/useSignatureInsights.js
Comment (FrederikBolding) [inline ui/hooks/useSignatureInsights.js:1]: Should be put in the snaps folder

File: ui/pages/confirm-signature-request/index.js (hunk 4 above)
Comment (hmalik88) [reply ui/pages/confirm-signature-request/index.js:83]: lol yeah woops

File: ui/components/app/snaps/snap-delineator/index.scss [hunk 6]
```diff
@@ -8,8 +8,13 @@
     &__text {
       white-space: nowrap;
       text-overflow: ellipsis;
+      width: calc(100% - 16px);
```
Comment (FrederikBolding) [inline ui/components/app/snaps/snap-delineator/index.scss:11]: Curious why this is suddenly necessary?

File: ui/helpers/utils/permission.js [hunk 7]
```diff
@@ -545,6 +545,43 @@ export const PERMISSION_DESCRIPTIONS = deepFreeze({
...
+        label: t('permission_signatureInsight'),
+        description: t('permission_signatureInsightDescription', [
+          getSnapNameComponent(targetSubjectMetadata),
+        ]),
+      },
+    ];
+
+    if (
+      isNonEmptyArray(permissionValue.caveats) &&
+      permissionValue.caveats[0].type === SnapCaveatType.SignatureOrigin &&
```
Comment (FrederikBolding) [inline ui/helpers/utils/permission.js:570]: I think we need to use `find` here soon. We might as well make that change now

File: ui/pages/confirm-signature-request/index.js [hunk 8]
```diff
@@ -190,6 +196,21 @@ const ConfirmTxScreen = ({ match }) => {
 
   const txData = useMemo(() => getTxData() || {}, [getTxData]);
 
+  ///: BEGIN:ONLY_INCLUDE_IF(build-flask)
+  const { data } = useSignatureInsights({ txData });
+
+  const warnings = data?.reduce((warningsArr, promise) => {
```
Comment (FrederikBolding) [inline ui/pages/confirm-signature-request/index.js:202]: Could the `useSignatureInsights` hook return this instead?

File: ui/hooks/snaps/useSignatureInsights.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    let cancelled = false;
+
+    async function fetchInsight() {
+      setLoading(true);
+
+      const {
+        msgParams: { from, data: msgData, signatureMethod, origin },
+      } = txData;
+
+      const shouldParse =
```
Comment (FrederikBolding) [inline ui/hooks/snaps/useSignatureInsights.js:28]: We should explain this edge-case with a comment.

File: ui/components/app/signature-request-original/signature-request-original.component.js [hunk 10]
```diff
@@ -348,19 +359,14 @@ export default class SignatureRequestOriginal extends Component {
         }}
         onSubmit={async () => {
           if (txData.type === MESSAGE_TYPE.ETH_SIGN) {
-            this.setState({ showSignatureRequestWarning: true });
-          } else {
-            ///: BEGIN:ONLY_INCLUDE_IF(build-mmi)
-            if (this.props.mmiOnSignCallback) {
```
Comment (FrederikBolding) [inline ui/components/app/signature-request-original/signature-request-original.component.js:354]: This logic seems to be removed? That seems bad

File: ui/components/app/signature-request/signature-request.js [hunk 11]
```diff
@@ -256,161 +269,191 @@ const SignatureRequest = ({ txData }) => {
   ///: END:ONLY_INCLUDE_IF
 
   return (
```
Comment (FrederikBolding) [inline ui/components/app/signature-request/signature-request.js:271]: We should probably get someone from the confirmations UX team to review these changes too.

And probably QA all the changed confirmations

File: ui/components/app/snaps/snap-delineator/snap-delineator.js [hunk 12]
```diff
@@ -83,6 +88,8 @@ export const SnapDelineator = ({
             marginLeft={1}
             marginTop={0}
             marginBottom={0}
+            marginRight={1}
+            display="block"
```
Comment (FrederikBolding) [inline ui/components/app/snaps/snap-delineator/snap-delineator.js:92]: Why is this suddenly necessary?

**Review by FrederikBolding — CHANGES_REQUESTED** [review]

File: ui/components/app/snaps/snap-delineator/index.scss (hunk 6 above)
Comment (hmalik88) [reply ui/components/app/snaps/snap-delineator/index.scss:11]: Had to do all of this to fix the header.

Comment (hmalik88) [reply ui/components/app/snaps/snap-delineator/index.scss:11]: We never tested the delineator title that would overflow

File: ui/components/app/snaps/snap-delineator/snap-delineator.js (hunk 12 above)
Comment (hmalik88) [reply ui/components/app/snaps/snap-delineator/snap-delineator.js:92]: Also for the header fix, needed margin and the element has to be "block" for text overflow to work.

File: ui/components/app/signature-request-original/signature-request-original.component.js (hunk 10 above)
Comment (hmalik88) [reply ui/components/app/signature-request-original/signature-request-original.component.js:354]: There's an `onSubmit` handler that I'm calling below that implements the same logic

**Review by FrederikBolding — CHANGES_REQUESTED** [review]
This implementation has problems with re-rendering causing a massive amount of requests to the snap. Blocking until we solve.

File: ui/hooks/snaps/useSignatureInsights.js [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+        setLoading(false);
+      }
+    }
+    if (txData) {
+      fetchInsight();
+    }
+    return () => {
+      cancelled = true;
+    };
+  }, [txData, snapIds, subjects]);
```
Comment (FrederikBolding) [inline ui/hooks/snaps/useSignatureInsights.js:109]:
```suggestion
  }, [txData, JSON.stringify(snapIds), subjects]);
```

Unfortunately we may need to do this

File: ui/hooks/snaps/useSignatureInsights.js [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+        setLoading(false);
+      }
+    }
+    if (txData) {
+      fetchInsight();
+    }
+    return () => {
+      cancelled = true;
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (FrederikBolding) [inline ui/hooks/snaps/useSignatureInsights.js:109]: Why is this needed?

Comment (hmalik88) [reply ui/hooks/snaps/useSignatureInsights.js:109]: it's complaining otherwise 
Screenshot 2024-02-07 at 7 08 23 AM

Comment (FrederikBolding) [reply ui/hooks/snaps/useSignatureInsights.js:109]: Curious, I didn't get that locally 🤔

File: ui/hooks/snaps/useSignatureInsights.js [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+        return warningsArr;
+      }, []);
+
+      if (!cancelled) {
+        setData(reformattedData);
+        setWarnings(insightWarnings);
+        setLoading(false);
+      }
+    }
+    if (txData) {
```
Comment (FrederikBolding) [inline ui/hooks/snaps/useSignatureInsights.js:103]: Let's check that it has keys since `txData` is sometimes an empty object.

Comment (FrederikBolding) [reply ui/hooks/snaps/useSignatureInsights.js:103]: We might wanna do this for transaction insights too 🤔

Comment (hmalik88) [reply ui/hooks/snaps/useSignatureInsights.js]: hmm, i'll do that in a different PR

Comment (hmalik88) [reply ui/hooks/snaps/useSignatureInsights.js]: for tx insights

**Review by FrederikBolding — DISMISSED** [review]
Code and performance LGTM. I'm not entirely sure about the confirmation pages changes, so we should get those looked at.

Let's have:
- Bowen do QA
- Erik do design review
- Someone from confirmations review and QA

File: ui/components/app/signature-request-original/signature-request-original.component.js [hunk 16]
```diff
@@ -400,10 +409,40 @@ export default class SignatureRequestOriginal extends Component {
...
+            origin={txData.msgParams.origin}
+            onCancel={() => {
+              this.setState({ showSignatureInsights: false });
+            }}
+            onSubmit={async () => {
+              await this.onSubmit();
+              this.setState({ showSignatureInsights: false });
+            }}
+          />
+        )}
```
Comment (jpuri) [inline ui/components/app/signature-request-original/signature-request-original.component.js:442]: Logic around setting `showSignatureInsights` can be moved into `InsightWarnings` or another component wrapping `InsightWarnings`.

This will avoid adding this state handling to all signature request pages individually.

File: ui/components/app/signature-request/signature-request.js [hunk 17]
```diff
@@ -234,160 +247,190 @@ const SignatureRequest = ({ txData }) => {
...
+          messageIsScrollable={messageIsScrollable}
+          primaryType={primaryType}
+        />
+        <Footer
+          cancelAction={onCancel}
+          signAction={() => {
+            ///: BEGIN:ONLY_INCLUDE_IF(build-flask)
+            if (warnings?.length >= 1) {
+              return setIsShowingSigInsightWarnings(true);
+            }
```
Comment (jpuri) [inline ui/components/app/signature-request/signature-request.js:375]: After signing page goes away, just curious is above code needed ?

File: ui/components/app/signature-request-original/signature-request-original.component.js (hunk 16 above)
Comment (hmalik88) [reply ui/components/app/signature-request-original/signature-request-original.component.js:442]: I had decided to keep logic at the signature request page levels because some of the logic is dependent on other state such as `showSignatureRequestWarning` and didn't want to introduce more complexity.

**Review by jpuri — DISMISSED** [review]

**Review by GuillaumeRx — DISMISSED** [review]
LGTM

(NIT: I wonder if we could extract the insight warning logic in a hook to avoid code duplication. It seems like a lot of code is duplicated across all the different signature screens we have)

**Review by GuillaumeRx — APPROVED** [review]

**Review by jpuri — APPROVED** [review]

**Comment by FrederikBolding** [comment]
@seaona 

For 1 I believe this is simply a timing issue, we don't currently block the UI from proceeding while waiting for snaps to respond. We could if that is preferable, but currently don't.

For 2 I don't think this is a bug in the implementation but just how the example snap works, without a domain the example snap does not have any insight to show: https://github.com/MetaMask/snaps/blob/main/packages/examples/packages/signature-insights/src/index.ts#L84-L109

**Review by FrederikBolding — APPROVED** [review]

**Comment by hmalik88** [comment]
Merging with acknowledgement that the first point raised by @seaona is a prod blocker for transaction insights and signature insights. The UI issues will be addressed in a forthcoming PR. The inconsistency in the snap export event is due to throttling that we will remove for certain exports in another PR: https://github.com/MetaMask/metamask-extension/issues/22917
